velox/functions/sparksql/Hash.cpp

```diff
@@ -32,16 +32,22 @@ void applyWithType(
     std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
     exec::EvalCtx& context,
     VectorPtr& resultRef) {
-  constexpr SeedType kSeed = 42;
+  SeedType seed = 42;
+  auto hashIdx = 0;
+  if (args[0]->isConstantEncoding()) {
```

Comment:
I'm puzzled by this logic. It seems like we interpret first arg as 'seed' if it is constant encoded and we use 42 for seed otherwise. However, encoding of the argument is an implementation detail and same query can end up having this argument encoded as flat or constant.

The function signature seems ambiguous to me as it seems impossible to reliably determine whether first argument is a seed or not.

Reply:
Yes. If the first argument is constant, it must be the seed. Or can we remove the default seed value to force specifying a seed value in the first argument?

Reply:
Here's the description of XxHash64 expression https://github.com/apache/spark/blob/3a9185964a0de3c720a6b77d38a446258b73468e/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/hash.scala#L648-L657
And here's the document for xxhash64 function in sparksql, which is mapped to the XxHash64 expression
https://github.com/apache/spark/blob/f7e8da3fb6dba44ab339a095c46cc8872d83b741/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L3232-L3239

As a sparksql function, there's no seed argument and it will always use default seed 42. But in Spark applications, it can have a different seed. e.g. In UT https://github.com/apache/spark/blob/f7e8da3fb6dba44ab339a095c46cc8872d83b741/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/HashExpressionsSuite.scala#L683-L708

Reply:
We can't pass these UTs using seed other than 42 in Gluten. Because the function is offloaded with the seed param, but there's nowhere to accept the param in Velox.

Reply:
Looking at 

```
case class XxHash64(children: Seq[Expression], seed: Long)
```

It seems that the signature of this function is (array(any), bigint), e.g. first argument is an array of arbitrary values and second argument is a bigint seed. Is this so?

In Velox, the signature of hash function is (any,...) which is different from above. 

One option is create 2 different functions: f1(any,...), f2(bigint, any,...) and map Spark expressions as appropriate. This way f2 would know for sure that first argument is the seed.